enclave/core/linux/threadlocal.c

```diff
@@ -124,7 +125,56 @@
  * calls to __tls_get_addr and to reduce thread-local variable access to simple
  * FS register offsets. To handle dynamically loaded modules we need to
  * implement __tls_get_addr.
- * For complete reference see: Elf Handling for Thread-Local Storage
+ * For complete reference see: Elf Handling for Thread-Local Storage.
+ *
+ * ****************************************************************************
+ * Exported thread-locals and shared-libraries:
+ *
+ * When thread-local variables are exported (via visibility=default), then the
+ * linker does not optimize the access of a thread-local variable to a constant
+ * offset from the FS register. Instead, for each thread-local variable,
+ * another variable is introduced varname@tpoff that contains the offset for
+ * the thread local variable. The offset value for varname@tpoff is expected to
+ * be filled up by the dynamic linker/loader.
+ * Consider,
+ *         __attribute__((visibility=default)) __thread int x;
+ *         int foo() { return x; }
+ *
+ * This results in the following code to access x
+ *     foo:
+ *         ...
+ *         mov %FS:0, %rax            // Fetch the address of the FS segment
+ *         add x@tpoff(%rip), %rax    // Fetch offset for x from the
+ *                                    // relocation entry x@tpoff and add it to
+ *                                    // FS
+ *         mov (%rax), rax            // Fetch value of x
+ *
+ * Each @tpoff variable results in a relocation entry of the type
+ * R_X86_64_TPOFF64. The relocation entry contains enough information to fill
+ * the value of the @tpoff variable.
+ * R_X86_64_TPOFF64 relocation info contains the following fields:
+ *
+ *    r_offset = relative address of the corresponding tpoff variable
+ *    r_info.relocation_type = R_X86_64_TPOFF64
+ *    r_info.symbol = index of symbol in the .dynsym section.
+ *    r_addend = 0
+ *
+ * The value (st_value) of the symbol in the .dynsym section contains the offset
+ * to the thread-local variable from the *start* of the thread-local section.
+ * Thus:
+ *    &x = tls-start + symbol sh_value.
+ *
+ * Since the compiler emits code relative to the end of the section (i.e using
+ * FS), the tpoff is computed via the formula:
+ *     tpoff = FS - (tls-start + sh-value).
+ *
+ * Thus, performing relocations for thread-local variables involves setting the
+ * value of the corresponding tpoff variables to the offset from the FS register
+ * value.
+ *
+ * To avoid looking up symbols within the enclave (symbols are not availabe)
```

Comment:
Very nice documentation. Curious if this is same for clang and gcc.

Reply:
Yes. Thread-local behavior is determined by ABI. 
The linker may or may not optimize all thread-local access to constant offsets; but where it can't it will use R_X86_64_TPOFF64 relocation.
Note: We don't support all kinds of thread-local relocations. Only the ones that are useful for an enclave context.